### PR TITLE
Add encryption support for Kinesis Firehose, S3 Bucket, and SNS topic

### PR DESCRIPTION
**Pull request by steveporcello**
## Summary

Customer has requirements to encrypt all AWS resources where data is stored at rest. This code enables encryption (by default) on resources created for the EKS audit log capabilities of Lacework: the Kinesis Firehose, S3 bucket, and SNS topic. Resources can have encryption enabled/disabled individually. By default this code will create a new KMS key to encrypt these resources, however a user can specify an existing key ARN to encrypt the resources as well.

## How did you test this change?

Multiple tests were performed: 1) Enabled encryption on all resources (Firehose,S3,SNS) with created KMS key 2) enable encryption only on certain combinations (Firehose+SNS, Firehose+S3, S3+SNS, etc) 3) disable all encryption 4) Enabled encryption on all resources with existing KMS key (with the right key policy permissions).

## Issue

https://lacework.atlassian.net/browse/ALLY-1077

### CODE REVIEW

**Comment by steveporcello** [comment]
Updated code to enable KMS key rotation

File: main.tf [hunk 1]
```diff
@@ -371,4 +417,98 @@ resource "lacework_integration_aws_eks_audit_log" "data_export" {
   depends_on = [time_sleep.wait_time_cw]
 }
 
+data "aws_iam_policy_document" "kms_key_policy" {
```
Comment (rmoles) [inline main.tf]: I wonder if this should live above `lacework_eks_kms_key` where it is used? Generally the last lines in our Terraform modules is the lacework api call.

**Comment by rmoles** [comment]
@lwsporcello Out of interest do we have a similar ticket to implement KMS via CloudFormation and a follow up ticket to update the customer facing docs?

File: main.tf [hunk 2]
```diff
@@ -1,28 +1,52 @@
...
+  sns_name                            = "${var.prefix}${random_id.uniq.hex}"
+  firehose_iam_role_name              = "${var.prefix}-fh-${random_id.uniq.hex}"
+  firehose_policy_name                = "${var.prefix}-fh-policy-${random_id.uniq.hex}"
+  firehose_delivery_stream_name       = "${var.prefix}${random_id.uniq.hex}"
+  eks_cw_iam_role_name                = "${var.prefix}-cw-${random_id.uniq.hex}"
+  cw_iam_policy_name                  = "${var.prefix}-cw-policy-${random_id.uniq.hex}"
+  cloudwatch_permission_resources     = "arn:aws:logs:*:${data.aws_caller_identity.current.account_id}:log-group:/aws/eks/*:*"
+  log_regions                         = [for region in var.cloudwatch_regions : "logs.${region}.amazonaws.com"]
+  cluster_names                       = var.no_cw_subscription_filter ? [] : var.cluster_names
+  create_kms_key                      = ((var.bucket_encryption_enabled && length(var.bucket_sse_key_arn) == 0) || (var.sns_topic_encryption_enabled && length(var.sns_topic_encryption_key_arn) == 0) || (var.kinesis_firehose_encryption_enabled && length(var.bucket_sse_key_arn) == 0)) ? 1 : 0
```
Comment (rmoles) [inline main.tf]: Nit: It might be nice to add a comment explaining the logic here. It's hard to parse at a glance and would be helpful for others in the future.

**Review by dmurray-lacework — APPROVED** [review]
Approved assuming @rmoles comments are addressed

**Comment by rmoles** [comment]
Make it so!

**Comment by dmurray-lacework** [comment]
make it so

**Comment by dmurray-lacework** [comment]
make it so please

**Comment by rmoles** [comment]
@lwsporcello would you mind adding a few examples to the `examples` folder?
Perhaps encryption enabled on all resources (Firehose,S3,SNS) with created KMS key & Enabled encryption on all resources with existing KMS key (with the right key policy permissions) would be sufficient

File: main.tf [hunk 3]
```diff
@@ -371,4 +417,98 @@ resource "lacework_integration_aws_eks_audit_log" "data_export" {
   depends_on = [time_sleep.wait_time_cw]
 }
 
+data "aws_iam_policy_document" "kms_key_policy" {
+  version = "2012-10-17"
+
+  statement {
+    sid    = "Enable IAM User Permissions"
```
Comment (rmoles) [inline main.tf]: It would be nice to be a bit more descriptive with the SID.

**Comment by steveporcello** [comment]
Accidentally closed the PR when doing a rebase, re-opening it with the changes suggested by Ross

**Comment by afiune** [comment]
Make it soooooo

File: main.tf (hunk 1 above)
Comment (afiune) [reply main.tf]: I didn't understand this comment 🤔

File: main.tf (hunk 3 above)
Comment (afiune) [reply main.tf]: All of this code is coming from our CloudTrail module 😜 https://github.com/lacework/terraform-aws-cloudtrail/blob/main/main.tf#L144

We probably need to do a sanity check there 👆🏽 too

**Review by afiune — APPROVED** [review]
tenor-222215594


I am going to merge this and create a PR to add the examples to the CI script since this PR has been there since a few days.

Thank you @lwsporcello for this feature!
